source/MicroBitPowerManager.cpp

```diff
@@ -370,40 +427,217 @@ void MicroBitPowerManager::deepSleep(uint32_t milliSeconds)
+            return DEVICE_INVALID_STATE;
+        }
+    }
+    else
+    {
+        if ( wakeUpSources)
+        {
+            manageSleepData data;
+            CodalComponent::manageAllSleep( manageSleepCountWakeUps, &data);
+            if ( data.count == 0)
+            {
+                DMESG( "deepSleep: no wake up sources");
+                return DEVICE_INVALID_STATE;
+            }
+        }
+        else
+        {
+            if ( wakeUpPin == NULL)
+            {
+                DMESG( "deepSleep: no wake up pin");
+                return DEVICE_INVALID_STATE;
+            }
+        }
+    }
+
     // Configure for sleep mode
-    setSleepMode(true);
 
-    // Wait a little while to ensure all hardware and peripherals have reacted to the change of power mode.
-    //target_wait(10);
+    setPowerLED( true /*doSleep*/);
 
-    // Ensure the requested pin into digital input mode. 
-    pin.getDigitalValue();
+    // Update peripheral drivers
+    CodalComponent::manageAllSleep( wakeUpSources ? manageSleepBeginWithWakeUps : manageSleepBegin, NULL);
+
+    CODAL_TIMESTAMP tickStart;
+    CODAL_TIMESTAMP timeStart = system_timer_deepsleep_begin( &tickStart);
+
+    int      channel      = 2;      //System timer uses period = 0, event = 1 and capture = 3
+    uint32_t saveCompare  = sysTimer->timer->CC[channel];
+    uint32_t saveIntenset = sysTimer->timer->INTENSET;
+
+    sysTimer->timer->INTENCLR = sysTimer->timer->INTENSET;
+
+    void (*sysTimerIRQ) (uint16_t channel_bitmsk) = sysTimer->timer_pointer;
+    sysTimer->setIRQ( deepSleepTimerIRQ);
+    timer_irq_channels = 0;
+
+//    ManagedString dbg = "IRQ  ";
+//    for ( int irq = Reset_IRQn; irq <= SPIM3_IRQn; irq++)
+//    {
+//      if ( NVIC_GetEnableIRQ( (IRQn_Type) irq))
+//      {
+//        dbg = dbg + ManagedString( (int) irq);
+//        dbg = dbg + " ";
+//      }
+//    }
+//    DMESG( dbg.toCharArray());
+//
+//    //These may be enabled: RADIO_IRQn, UARTE0_UART0_IRQn
+//    //With BLE, also: POWER_CLOCK_IRQn, RTC0_IRQn, RTC1_IRQn, SWI2_EGU2_IRQn, SWI5_EGU5_IRQn, MWU_IRQn
+
+    uint32_t usPerTick   = 1;
+    uint32_t ticksPerMS  = 1000;
+    uint32_t ticksMax    = 0xFFFFFFFFul - ticksPerMS * 1000; // approx 71min
+    //ticksMax = 2000000;
+
+    if ( !wakeUpSources)
+    {
+        if ( wakeUpPin)
+        {
+            // Ensure the requested pin into digital input mode. 
+            wakeUpPin->getDigitalValue();
 
-    // Enable an interrupt on the requested pin or an interrupt from the KL27.
-    pin.setDetect(pin.getPolarity() ? GPIO_PIN_CNF_SENSE_High : GPIO_PIN_CNF_SENSE_Low);
+            // Enable an interrupt on the requested pin or an interrupt from the KL27.
+            wakeUpPin->setDetect(wakeUpPin->getPolarity() ? GPIO_PIN_CNF_SENSE_High : GPIO_PIN_CNF_SENSE_Low);
+        }
+    }
 
     // Enable wakeup from the the KL27 interrupt line.
     io.irq1.setDetect(GPIO_PIN_CNF_SENSE_Low);
+    NVIC_EnableIRQ(GPIOTE_IRQn);
```

Comment:
hi @martinwork 

GPIOTE is used in lots of places... Why do we need this? I'm a bit concerned there may be a lot of IRQs from light sensor, display....

Reply:
While not at all sure, I thought I confirmed that GPIOTE_IRQn seems to be necessary to get wake up IRQs from pins and buttons. This line is a bit of a hang over from a previous attempt which disabled then enabled it on the way in and out, but I thought as it's needed for pins, it would be needed for io.irq1. How do I send irq1 low?

Reply:
If it is required, all those places will have to disable their use of  GPIOTE_IRQn, and if it's desirable to disable GPIOTE_IRQn when it's not needed we'll have to ask them all if they need it. I believe deepSleep(MicroBitPin &pin) was running with GPIOTE_IRQn enabled.